File: Balanced Binary Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if(!root)
            return true;
        int diff = abs(maxHeight(root->left) - maxHeight(root->right));
        if(diff > 1)
            return false;
        return isBalanced(root->left) && isBalanced(root->right);
    }
    
    int maxHeight(TreeNode *root)
    {
        if(!root)
            return 0;
        if(!root->left && !root->right)
            return 1;
        return 1 + max(maxHeight(root->left), maxHeight(root->right));
    }


};

File: python/Balanced Binary Tree.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isBalanced(self, root):
        if root is None:
            return True
        diff = abs(self.maxHeight(root.left) - self.maxHeight(root.right))
        if diff > 1:
            return False

        return self.isBalanced(root.left) and self.isBalanced(root.right)

    def maxHeight(self, node):
        if node is None:
            return 0

        return 1 + max(self.maxHeight(node.left), self.maxHeight(node.right))